### ISSUE TITLE
Error: ENOENT, when deleting empty folder which used to contain a file

### ISSUE THREAD

**Issue by mozeryansky** [ember-cli/ember-cli#5297]
An error occurs when I delete an empty folder when a file used to be in it.
This does not crash ember-cli, just displays an error.

I am on version 2.2.0-beta.2.
Below is the console output from running `ember s`.

1. Create File
```
file added templates/untitled file

Build successful - 219ms.
```
2. Create folder
```
file added templates/untitled folder

Build successful - 211ms.
```
3. Move file into folder
```
file deleted templates/untitled file
file added templates/untitled folder/untitled file

Build successful - 242ms.    
```
4. Move file out of folder
```
file deleted templates/untitled folder/untitled file
file added templates/untitled file

Build successful - 237ms.
```
5. Delete now empty folder
```
file deleted templates/untitled folder
ENOENT: no such file or directory, rmdir '/Users/michael/Sites/super-rentals/tmp/template_compiler-output_path-nC2C6Cdc.tmp/super-rentals/templates/untitled folder'
Error: ENOENT: no such file or directory, rmdir '/Users/michael/Sites/super-rentals/tmp/template_compiler-output_path-nC2C6Cdc.tmp/super-rentals/templates/untitled folder'
    at Error (native)
    at Object.fs.rmdirSync (fs.js:806:18)
    at TemplateCompiler.<anonymous> (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/index.js:83:32)
    at /Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/promise-map-series/index.js:11:14
    at lib$rsvp$$internal$$tryCatch (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/rsvp/dist/rsvp.js:493:16)
    at lib$rsvp$$internal$$invokeCallback (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/rsvp/dist/rsvp.js:505:17)
    at /Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/rsvp/dist/rsvp.js:1001:13
    at lib$rsvp$asap$$flush (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/rsvp/dist/rsvp.js:1198:9)
    at nextTickCallbackWith0Args (node.js:433:9)
    at process._tickCallback (node.js:362:13)
```

**Comment by stefanpenner** [comment]
unfortunately I was unable to replicate this issue using 2.2.0-beta.3 in a fresh app.

my attempt (in-case I missed a step):

```
file added the-file

Build successful - 131ms.

file added the-folder

Build successful - 112ms.

file deleted the-file
file added the-folder/the-file

Build successful - 89ms.

file deleted the-folder/the-file
file added the-file

Build successful - 127ms.

file deleted the-folder

Build successful - 107ms.

```

**Comment by stefanpenner** [comment]
a replicated it, it must be in a nested directory.

**Comment by stefanpenner** [comment]
ya, its totally unlinking the directory

```
file deleted templates/the-folder/the-file
file added templates/the-file
create /Users/stefanpenner/tmp/deleting-empty-folder/tmp/broccoli_merge_trees-input_base_path-PnKnTw46.tmp/14/the-file /Users/stefanpenner/tmp/deleting-empty-folder/tmp/broccoli_merge_trees-output_path-l6SccJzd.tmp/the-file
unlinkdir undefined /Users/stefanpenner/tmp/deleting-empty-folder/tmp/broccoli_merge_trees-output_path-5O0bkeat.tmp/deleting-empty-folder/templates/the-folder
create /Users/stefanpenner/tmp/deleting-empty-folder/tmp/broccoli_merge_trees-input_base_path-XrUW2i1W.tmp/1/deleting-empty-folder/templates/the-file /Users/stefanpenner/tmp/deleting-empty-folder/tmp/broccoli_merge_trees-output_path-5O0bkeat.tmp/deleting-empty-folder/templates/the-file
```

Need to pin-point where it leaves us.

**Comment by stefanpenner** [comment]
looks like the problem can be reduced to

starting state:
```
app/templates/the-folder/the-file
```

single failure operation
```
mv app/templates/the-folder/the-file app/templates/the-file
```

**Comment by stefanpenner** [comment]
Looks like their are some additional issues relative to empty directories (as I'm playing around with them).

Looks also like the issue is between the templates TreeMerger (which has the folder in its output) and the preprocessedApp & templates treeMerge (which appears to be missing the folder in its input trees)

**Comment by mozeryansky** [comment]
I was testing a bit too, and reversed the mv operation:
`mv app/templates/the-file2 app/templates/the-folder2/the-file2`

and got this error:
```
file deleted templates/the-file2
file added templates/the-folder2/the-file2
ENOENT: no such file or directory, symlink '/Users/michael/Sites/super-rentals/tmp/template_compiler-input_base_path-jfASOBi6.tmp/0/super-rentals/templates/the-folder2/the-file2' -> '/Users/michael/Sites/super-rentals/tmp/template_compiler-output_path-dJIphaai.tmp/super-rentals/templates/the-folder2/the-file2'
Error: ENOENT: no such file or directory, symlink '/Users/michael/Sites/super-rentals/tmp/template_compiler-input_base_path-jfASOBi6.tmp/0/super-rentals/templates/the-folder2/the-file2' -> '/Users/michael/Sites/super-rentals/tmp/template_compiler-output_path-dJIphaai.tmp/super-rentals/templates/the-folder2/the-file2'
    at Error (native)
    at Object.fs.symlinkSync (fs.js:945:18)
    at symlink (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/symlink-or-copy/index.js:82:14)
    at symlinkOrCopySync (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/symlink-or-copy/index.js:58:5)
    at TemplateCompiler.Filter._handleFile (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/index.js:101:12)
    at TemplateCompiler.<anonymous> (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/index.js:88:21)
    at /Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/promise-map-series/index.js:11:14
    at lib$rsvp$$internal$$tryCatch (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/rsvp/dist/rsvp.js:493:16)
    at lib$rsvp$$internal$$invokeCallback (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/rsvp/dist/rsvp.js:505:17)
    at lib$rsvp$$internal$$publish (/Users/michael/Sites/super-rentals/node_modules/ember-cli-htmlbars/node_modules/broccoli-persistent-filter/node_modules/rsvp/dist/rsvp.js:476:11)
```

**Comment by stefanpenner** [comment]
I'll likely dig into this tomorrow. I have an idea, but need to first validate it.

**Comment by mozeryansky** [comment]
I looked into this for a long time, and I have no idea what I'm looking for. Good luck.

The best I could determine is the error is thrown from `readAndReturnNodeFor` in builder.js.

**Comment by stefanpenner** [comment]
I believe the issue is a mixture of broccoli filters, their exists a scenario were either:

1. the directory is in-correctly pruned
2. the directory is correctly pruned, but do to a symlink optimization a subsequent filter expects to also prune the given directory.

I have solutions for both cases, but  the best bet, is for someone (likely me later this week) to come up with a minimal reproduction. That way we can be sure of the issue, before applying the solution.

**Comment by sly7-7** [comment]
I have the same error when deleting a single in a folder. Adding a new file in the same folder, and ember-cli is building again.

**Comment by sly7-7** [comment]
@stefanpenner I'm not sure how related it is, but in my use case, when deleting a non js file, ember-cli is still happy, but when deleting a js file, it crashes. Seems like it's related to jshinter broccoli filter.
```
Error: ENOENT: no such file or directory, unlink '/Users/sma01/work/svn/webapps/esatis/branches/materialized/frontend/tmp/jshinter-cache_path-ScXc4Vmc.tmp/helpers/is-finite.js'
):
  Error: ENOENT: no such file or directory, unlink 'frontend/tmp/jshinter-cache_path-ScXc4Vmc.tmp/helpers/is-finite.js'
      at Error (native)
      at Object.fs.unlinkSync (fs.js:981:18)
      at JSHinter.<anonymous> (frontend/node_modules/broccoli-filter/index.js:60:8)
      at Array.forEach (native)
      at JSHinter.build (frontend/node_modules/broccoli-filter/index.js:59:35)
      at JSHinter.build (frontend/node_modules/broccoli-jshint/index.js:41:33)
      at frontend/node_modules/broccoli-plugin/read_compat.js:61:34
```

**Comment by stefanpenner** [comment]
@hjdivad you and me should take a look at this one soon.

**Comment by hjdivad** [comment]
@mozeryansky if you have a moment it would be helpful to check whether #5965 fixes this problem in your app.  It looks very likely to.

**Comment by hjdivad** [comment]
trolled by commit message; still waiting on #5969

**Comment by hjdivad** [comment]
nevermind, @stefanpenner corrected me, npm will get the right dependencies which *are* released.

**Comment by stefanpenner** [comment]
should be fixed with broccoli-persistent-filter v1.2.6, broccoli-merge-tree v1.1.2, broccoli-funnel 1.0.3 as tree-sync 1.1.2  and fs-tree-diff 0.5.0 which all have been released, but some applications may require dependency updates.

**Comment by hjdivad** [comment]
also walk-sync 0.2.7
